add propeller docs (#34)
equinox-project/equinox-js: docs/docs/propeller/index.md

---
sidebar_position: 4
---

# Propeller

Propeller is EquinoxJS's "[propulsion](https://github.com/jet/propulsion)
lite". It's a set of abstractions for building reactive event processing
pipelines catering specifically for event sourced reactions from Equinox's
stores.

## StreamsSink

The `StreamsSink` receives batches of events from a source and intelligently
manages handling them.

- **Stream management**: maintains an ordered queue of incoming streams, ensuring

>   that events are processed in an orderly manner.

Reviewer: Probably merits an individual point that the goal is to close batches and checkpoint them, and that this is achieved without needing to stop the world for the checkpointing operation to complete before we move to the next one.

```suggestion
  that reactions processing aligns with the ordering of the triggering events; the aim being to be able to checkpoint attained progress ('tick off batches') as soon as possible.
```

- **Concurrency control**: To ensure optimal performance, StreamsSink restricts the
  number of streams and event batches that can be processed at once. It
  efficiently handles the inflow of new event batches, ensuring the system
  doesn't get overwhelmed.
- **Batch merging**: intelligently merges new events with existing streams. This
  allows for a more streamlined event processing, ensuring that events from

>   multiple batches related to the same stream are handled together.

Reviewer: >This
  allows for a more streamlined event processing, ensuring that events from
  multiple batches related to the same stream are handled together.
  
Where multiple batches include events from the same stream, these are supplied to the handler as a single sequence; this allows handler throughput to to be optimized (note the sink also guarantees that there will never be more than one handler processing events for a given stream)

(and if that last part is not true yet, it needs to be; probably worth making it a separate point so its easier to read and stands out)

## TailingFeedSource

`TailingFeedSource` continuously crawls a feed, submitting batches of events to
a `Sink`.

> - **Crawling**: Continuously crawls a given feed. It pauses momentarily between crawls once

Reviewer: ```suggestion
- **Crawling**: Continuously crawls a given feed forward from a checkpointed position (or the start of the feed if no batches have yet completed processing). It pauses momentarily between crawls once
```

  it's reached the tail of the feed.

> - **Checkpointing**: Regularly updates the checkpoint position to match the latest

Reviewer: ```suggestion
- **Checkpointing**: Regularly updates the checkpoint position to reflect the latest completed
```
should mention something about this being async to the processing and deduplicated if there are lots of batches being processed very quickly

>   processed batch. Flushes the checkpoint when interrupted to minimise reduntant

Reviewer: ```suggestion
  processed batch. Flushes the checkpoint when interrupted to minimise redundant
```

---
sidebar_position: 4
---

# Propeller

Propeller is EquinoxJS's "[propulsion](https://github.com/jet/propulsion)

> lite". It's a set of abstractions for building reactive event processing

Reviewer: the abstractions are not the sell really - its the fact that there are impls, and that they have a ecent structure that will yield you a testable/separated handler not mangled in store specific bpoilerplate.

It's a set of components for building ...
The primary purpose is to manage reactions to events emanating from an event store's feed.
The components all follow common high level patterns, with specific treatments and options to allow you to manage specific details of the actual store you're using.
The high level structure is also designed to work well in terms of being able to test handlers in isolation, or deal with similar needs, such as continually ingesting data from third party event feeds.

(last point is probably not intro material)

Author: You are absolutely correct. It's just so much easier to write about the abstractions :(

pipelines catering specifically for event sourced reactions from Equinox's
stores.

## StreamsSink

The `StreamsSink` receives batches of events from a source and intelligently
manages handling them.

- **Stream management**: maintains an ordered queue of incoming streams,
  ensuring that reactions processing aligns with the ordering of the triggering
  events; the aim being to be able to checkpoint attained progress ('tick off
  batches') as soon as possible.
- **Concurrent stream handling**: To ensure optimal performance, `StreamsSink`

>   restricts the number of streams that are processing at any given moment to a

Reviewer: ->that are processing
for which handlers are processing

maybe 'in flight' ?

>   configurable limit. Additionally guarantees that each stream can have at most

Reviewer: This is not a lil bonus - its a central thing

>Additionally guarantees that each stream can have at most
  one handler actively processing.

Note, while handlers work concurrently, there's only ever a single handler in flight for any given stream; it gets passed the queue of all pending events for that stream 

(cant think of a good sentence to convey "and you can do them sequentially, run them in parallel, do batch SQL", without contention knowing that there are not other events for that stream being handled at the same time)

  one handler actively processing.

> - **Batch merging**: intelligently merges new events with existing streams.

Reviewer: maybe this goes before the preceding point, leading into the fact that the handler gets the latest coalesced stream events from across batches

  This allows for a more streamlined event processing, ensuring that events
  from multiple batches related to the same stream are handled together.

## TailingFeedSource

`TailingFeedSource` continuously crawls a feed, submitting batches of events to
a `Sink`.

- **Crawling**: Continuously crawls a given feed forward from a checkpointed
  position (or the start of the feed if no batches have yet completed
  processing). It pauses momentarily between crawls once it's reached the tail
  of the feed.

> - **Checkpointing**: Regularly updates the checkpoint position to reflect the

Reviewer: regularly sounds like it's a pretty random sort of a contraption. Periodic?

Reviewer: maybe it should allude to or mention that the wait to checkpoint is not part of the processing time overall and/or that it dedups checkpoint writes if there are lots in a replay scenario?
